File: bff.php
#!/usr/bin/php
<?php

$args = $argv;
$gcc_args = array();

$O = array(
	'combine' => true
);

array_shift($args);

while ($args[0][0] == '-') {
	if ($args[0] == 'nc') {
		$O['combine'] = false;
	} else {
		$gcc_args[] = array_shift($args);
	}
}

//$file = "php://stdin";
//if (count($args)) {
	$file = $args[0];
//}
//$stream = fopen($file, 'r');

$o_file = str_replace('.bf', '', $file);
$c_file = $o_file . '.c';

$code = file_get_contents($file);

init();
yyinit();
yyfill_str($code);

ob_start();

echo
"#define MOVL1     ptr--;
#define MOVL(__v) ptr -= (__v);
#define MOVR1     ptr++;
#define MOVR(__v) ptr += (__v);
#define INC1      (*ptr)++;
#define INC(__v)  (*ptr) += (__v);
#define DEC1      (*ptr)--;
#define DEC(__v)  (*ptr) -= (__v);
#define GET       (*ptr) = getchar();
#define PUT       putchar(*ptr);
#define LOOP      while (*ptr)

char st[32768];
int main(void) {
	char *ptr = st;\n\n";

yylex();

echo "}";

$src = ob_get_clean();
file_put_contents($c_file, $src);

$cmd = "gcc " . implode(" ", $gcc_args) . " -o " . escapeshellarg($o_file) . " " . escapeshellarg($c_file);

shell_exec($cmd);





function yylex() {
	global $yytext, $yych, $yycursor, $yylimit;
	global $O;

	while ($yych = yynext()) {
		if ($yych == '/' && yypeek('/')) {
			consumeToEOL();
		} else if ($yych == '[') {
			emit('LOOP {');
			indent(1);
		} else if ($yych == ']') {
			indent(-1);
			emit('}');
		} else if ($yych == ',') {
			emit(GET);
		} else if ($yych == '.') {
			emit(PUT);
		} else if ($yych == '#') {
			/* debug */
		} else {
			$count = $O['combine'] ? consumeChar($yych) : 1;
			$cmd = false;

			switch ($yych) {
				case '+': $cmd = INC;  break;
				case '-': $cmd = DEC;  break;
				case '<': $cmd = MOVL; break;
				case '>': $cmd = MOVR; break;
			}

			if ($cmd) {
				if ($count > 1) {
				emit($cmd, "(", $count, ")");
				} else {
					emit($cmd . '1');
				}
			}
		}
	}
}



function init() {
	global $_indent, $_indent_text;

	$_indent = 1;
	$_indent_text = "\t";
}

function yyinit() {
	global $yytext, $yych, $yycursor, $yylimit;

	$yytext = '';
	$yych = '';
	$yycursor = -1;
	$yylimit = 0;
}

function yyfill_str($str) {
	global $yytext, $yych, $yycursor, $yylimit;

	$yytext = $str;
	$yylimit = strlen($yytext);
}


function emit($args) {
	echo indent();

	if (!is_array($args)) {
		$args = func_get_args();
	}

	for ($i = 0, $l = count($args); $l > $i; ++$i) {
		echo $args[$i];
	}

	echo "\n";
}

function indent($delta = null) {
	global $_indent, $_indent_text;

	$argc = func_num_args();
	if ($argc == 0) {
		return $_indent_text;

	} else if ($delta < 0 && $delta <= $_indent) {
		$_indent += $delta;
		$_indent_text = substr($_indent_text, 0, $_indent);

	} else if ($delta > 0) {
		$_indent += $delta;
		$_indent_text .= str_repeat("\t", $delta);
	}
}

function yynext($assert = null) {
	global $yytext, $yych, $yycursor, $yylimit;

	if ($yycursor + 1 < $yylimit) {
		if (is_string($assert)) {
			if ($assert == $yytext[$yycursor + 1]) {
				++$yycursor;
				return true;
			}
		} else {
			++$yycursor;
			$yych = $yytext[$yycursor];
			return $yych;
		}
	}

	return false;
}

function yypeek() {
	global $yytext, $yych, $yycursor, $yylimit;

	$argv = func_get_args();
	$argc = count($argv);

	$offset = 1;
	$assert = null;

	if ($argc == 2 || is_int($argv[0])) {
		$offset = array_shift($argv);
	}

	$assert = array_shift($argv);

	if (is_string($arg1)) {
		$assert = $arg1;
	} else if (!is_null($arg1)) {
		$offset = $arg1;
		if (is_string($arg2)) {
			$assert = $args;
		}
	}

	$cursor = $yycursor + $offset;
	if ($cursor < $yylimit) {
		$ch = $yytext[$cursor];
		if (!is_null($assert)) {
			$ch = $assert === $ch;
		}
		return $ch;
	}

	return false;
}

function consumeChar($char) {
	global $yycursor;
	$start = $yycursor;
	while (yynext($char));
	return $yycursor - $start + 1;
}

function consumeToEOL() {
	global $yycursor;
	$offset = 0;
	while (!yypeek(++$offset, "\n"));
	$yycursor += $offset;
}

File: bff.c
#include "bff.h"

#define yych (*yycursor)

char *filename = NULL, *cfilename = NULL, *ofilename = NULL;
bool write_to_file = false;
FILE *input, *output;

char *indent, *indent_start;
char *yytext, *yycursor, *yylimit;

int main(int argc, char **argv, char **env)
{
	if (argc < 2 || argv[1] == NULL) {
		input = stdin;
		cfilename = stralloc(L_tmpnam);
		tmpnam(cfilename);
	} else {
		write_to_file = true;
		filename = argv[1];
		populateFilenames();
		input = fopen(filename, "r");
	}

	init();

	output = fopen(cfilename, "w");

	file_get_contents(input, &yytext);
	yyfill_str(yytext);

	fputs(intro, output);

	yylex();

	fputc('}', output);

	fflush(output);
	fclose(output);

	fclose(input);

	call_gcc();

	xfree(yytext);
	xfree(cfilename);
	xfree(ofilename);

	return 0;
}

#define BUFFER_SIZE 65536

void file_get_contents(FILE *fd, char **data)
{
	*data = (char *) malloc(BUFFER_SIZE);
	size_t len = fread(*data, 1, BUFFER_SIZE, fd);
	fputs(*data, stdout);
	printf("len: %d\nlen2: %d\n", (int)strlen(*data), (int)len);
	/*char *buffer = (char *) malloc(BUFFER_SIZE),
		 *buffer_end = buffer;
	size_t read = 0, buffer_len = 0, buffer_size = 0;
	while (read = fread(buffer_end, 1, BUFFER_*/
}


void populateFilenames(void)
{
	size_t len = strlen(filename);
	char *tmp, *ext;

	ofilename = stralloc(len);
	cfilename = stralloc(len);

	tmp = ofilename;
	strcpy(tmp, filename);
	ext = tmp + len - 3;

	ext[2] = '\0';
	ext[1] = 'c';
	strcpy(cfilename, tmp);

	ext[0] = '\0';
}


void call_gcc(void)
{
	char cmd[1024];
	sprintf(cmd, "gcc -O3 -g -o %s %s", ofilename, cfilename);
	system(cmd);
}


void yylex() {
	char tmp[128],
		 *cmd = NULL,
		 *start = NULL;
	size_t count = 1;
	while (yynext())
	{
		switch (yych) {
			//case '/': if (yypeek() == '/' && yyscan('\n')) emit("/* comment detected */"); break;
			//case '#': emit("/* hash detected */"); break;

			case ',': emit("GET"); break;
			case '.': emit("PUT"); break;

					// indenting is backward
			case '[': emit("LOOP {"); indent--; break;
			case ']': indent++; emit("}"); break;

			case '+':
			case '-':
			case '<':
			case '>': {
				//char *start = yycursor;
				//yyscan(yych);
				//size_t count = yycursor - start + 1;

				switch (yych) {
					case '+': cmd = "INC";  break;
					case '-': cmd = "DEC";  break;
					case '<': cmd = "MOVL"; break;
					case '>': cmd = "MOVR"; break;
				}

				if (cmd != NULL) {
					//if (count > 1) {
						//sprintf(tmp, "%s(%d)", cmd, (int)count);
					//} else {
						sprintf(tmp, "%s1", cmd);
					//}
					emit(tmp);
				}

				break;
			 }
		}
	}
}


void init(void)
{
	indent_start = stralloc(127);
	memset(indent_start, '\t', 127);
	indent_start[127] = '\0';
	indent = indent_start + 126;
}


void yyfill_str(char *str)
{
	yytext = str;
	yycursor = yytext - 1;
	yylimit = yycursor + strlen(yytext);
}


void emit(char *str)
{
	fputs(indent, output);
	fputs(str, output);
	fputc('\n', output);
}


int yynext()
{
	if (yycursor + 1 < yylimit) {
		++yycursor;
		return 1;
	}
	return 0;
}


char yypeek()
{
	if (yycursor + 1 < yylimit) {
		return yycursor[1];
	}
	return 0;
}

int yyscan(char ch)
{
	while (yycursor + 1 < yylimit && ++yycursor && yych != ch);
	return yych == ch;
}
